man/man5/postfilter.conf.5

.TH postfilter.conf 5 "January 20, 2019" "" "Postfilter main configuration file"

.SH NAME
postfilter.conf - Postfilter main configuration file

.SH DESCRIPTION

postfilter.conf is the main postfilter configuration file. This file includes all the most important settings and
.B must be properly configured
before starting to use postfilter. The default values are set in order to provide the highest antispam and anti-abuse protection even if they 
are probably too restrictive for sites that use some sort of users authentication. Servers that offer unauthenticated access can use them as
they are.
.br
.B Beware:
configuration arguments are
.B case sensitive
so 'TRUE' and 'true' are 
.B not
the same. All postfilter configuration values 
.B must be
typed using only
.B minuscule letters.
In order to avoid strange and hard to predict behaviours, it's higly recommended to assign a value to all configuration directives that are stored 
inside
postfilter.conf. Deletion of configuration keys should be avoided. 

.SH List of all variables
.P
The following variables can be configured through postfilter.conf:
.P
.td

.B 01.
$config{'salt'}                            random string 
.br
.B 02.
$config{'server_status'}                   [ 'active' | 'closed' | 'disabled ]
.br
.B 03.
$config{'server_type'}                     [ 'public' | 'auth' | 'both' ]
.br
.B 04.
$config{'enable_domain_check'}             [ 'true' | 'false' ]
.br
.B 05.
$config{'default_action_on_accept'}        [ 'accept' | 'discard' | 'save' | 'reject' ]
.br
.B 06.
$config{'default_action_on_reject'}        [ 'accept' | 'discard' | 'save' | 'reject' ]
.br
.B 07.
$config{'public_user_id'}                  'RegEX'
.br
.B 08.
$config{'reject_on_badwords_error'}        [ 'true' | 'false' ]
.br
.B 09.
$config{'reject_on_banlist_error'}         [ 'true' | 'false' ]
.br
.B 10.
$config{'show_error_code'}                 [ 'true' | 'false' ]
.br
.B 11.
$config{'period'}                          Positive, integer, number of seconds
.br
.B 12.
$config{'short_period'}                    Positive, integer, number of seconds
.br
.B 13.
$config{'trash_period'}                    Positive, integer, number of seconds
.br
.B 14.
$config{'enable_mysql'}                    [ 'true' | 'false' ]
.br
.B 15.
$config{'check_groups_existence'}          [ 'true' | 'false' ]
.br
.B 16.
$config{'check_users'}                     [ 'true' | 'false' ]
.br
.B 17.
$config{'check_tor'}                       [ 'true' | 'false' ]
.br
.B 18.
$config{'tor_network'}                     [ 'allow' | 'reject' | 'mark' ]
.br
.B 19.
$config{'check_white-list'}                [ 'true' | 'false' ]
.br
.B 20.
$config{'check_custom'}                    [ 'true' | 'false' ]
.br
.B 21.
$config{'check_banlist'}                   [ 'true' | 'false' ]
.br
.B 22.
$config{'score_banlist'}                   [ integer number ]
.br
.B 23.
$config{'check_rbl'}                       [ 'true' | 'false' ]
.br
.B 24.
$config{'check_uribl'}                     [ 'true' | 'false' ]
.br
.B 25.
$config{'check_surbl'}                     [ 'true' | 'false' ]
.br
.B 26.
$config{'check_uriblcom'}                  [ 'true' | 'false' ]
.br
.B 27.
$config{'check_badwords'}                  [ 'true' | 'false' ]
.br
.B 28.
$config{'scan_body'}                       [ 'true' | 'false' ]
.br
.B 29.
$config{'max_score_on_body'}               Integer, positive, number
.br
.B 30.
$config{'scan_subject'}                    [ 'true' | 'false' ]
.br
.B 31.
$config{'max_score_on_subject'}            Integer, positive, number
.br
.B 32.
$config{'allow_control_cancel'}            [ 'true' | 'false' ]
.br
.B 33.
$config{'allow_supersedes'}                [ 'true' | 'false' ]
.br
.B 34.
$config{'allow_uuencode'}                  [ 'true' | 'false' ]
.br
.B 35.
$config{'allow_yenc'}                      [ 'true' | 'false' ]
.br
.B 36.
$config{'allow_mail_headers'}              [ 'true' | 'false' ]
.br
.B 37.
$config{'allow_html'}                      [ 'true' | 'false' ]
.br
.B 38.
$config{'legal_summary'}                   [ 'true' | 'false' ]
.br
.B 39.
$config{'check_distribution'}              [ 'true' | 'false' ]
.br
.B 40.
$config{'max_followup'}                    Positive, integer, number of allowed groups
.br
.B 41.
$config{'max_groups_difference'}           Positive integer number of groups
.br
.B 42.
$config{'max_crosspost'}                   Positive, integer, number of allowed groups
.br
.B 43.
$config{'max_fup_no_crsspt'}               Positive, integer, number of allowed groups
.br
.B 44.
$config{'maximum_multipost'}               an integer positive number
.br
.B 45.
$config{'md5_hash'}                        [ 'body' | 'basic' | 'complex' | 'all' ]
.br
.B 46.
$config{'max_body_size'}                   Positive, integer, number of bytes
.br
.B 47.
$config{'max_head_size'}                   Positive, integer, number of bytes
.br
.B 48.
$config{'max_total_size'}                  Positive, integer, number of bytes
.br
.B 49.
$config{'max_header_length'}               Positive, integer, number of bytes
.br
.B 50.
$config{'max_hierarchies_post'}            Positive, integer, number of hierarchies
.br
.B 51.
$config{'max_hierarchies_followup'}        Positive, integer, number of hierarchies
.br
.B 52.
$config{'max_quoted_ratio'}                Decimal number between 0 and 1
.br
.B 53.
$config{'max_blank_ratio'}                 Decimal number between 0 and 1
.br
.B 54.
$config{'max_empty_ratio'}                 Decimal number between 0 and 1
.br
.B 55.
$config{'max_grace_time'}                  Integer, positive, number of seconds
.br
.B 56.
$config{'too_old_limit'}                   Integer, positive, number of seconds
.br
.B 57.
$config{'max_line_length'}                 Integer, positive, number of characters
.br
.B 58.
$config{'force_default_organization'}      [ 'true' | 'false' ]
.br
.B 59.
$config{'delete_header_x-trace'}           [ 'true' | 'false' ]
.br
.B 60.
$config{'delete_header_sender'}            [ 'true' | 'false' | 'anon' ]
.br
.B 61.
$config{'delete_header_nntp-posting-host'} [ 'true' | 'false' | 'anon' ]
.br
.B 62.
$config{'delete_header_nntp-posting-date'} [ 'true' | 'false' ]
.br
.B 63.
$config{'delete_header_user-agent'}        [ 'true' | 'false' ]
.br
.B 64.
$config{'delete_header_x-no-archive'}      [ 'true' | 'false' ]
.br
.B 65.
$config{'delete_mail_headers'}             [ 'true' | 'false' ]
.br
.B 66.
$config{'delete_custom_headers'}           [ 'true' | 'false' ]
.br
.B 67.
$config{'include_new_headers'}             [ 'true' | 'false' ]
.br
.B 68.
$config{'force_valid_path'}                [ 'true' | 'false' ]
.br
.B 69.
$config{'add_cancel_lock'}                 [ 'true' | 'false' ]
.br
.B 70.
$config{'lock_method'}                     [ 'sha1' | 'md5' ]
.br
.SH OPTIONS
.P

.br
.B 1. $config{'salt'} -> random string 
.br
.P
A random string needed by postfilter in order to add entropy to NNTP-Posting-Host Md5 sign.
This random string should be longer than 6 characters and must be changed before starting to
use postfilter in a production host.

.br
.B 2. $config{'server_status'} -> [ 'active' | 'closed' | 'disabled ]
.br
.P
Set the server status. A value of 'active' means that all checks are enabled. This is the most
common configuration value and should almost always be kept. if this is set to  'closed', 
all messages will be rejected; a value of 'disabled' disable every kind of checks: all
messages will be always accepted. 'closed' is a fast way to make the server read only without
touching readers.conf. 'disabled' is dangerous because the server becomes completely open
and this makes simple large abuses.

.br
.B 3. $config{'server_type'} -> [ 'public' | 'auth' | 'both' ]
.br
.P
Postfilter is able to detect whether someone has the right to post an article using the client's
userid or IP address. If the server uses authentication, the users' access rights must be 
calculated using the client's userid. If the clients aren't authenticated, it's needed to use
the client's IP address because nnrpd assign to all the same userid.
If *all* users have got an userid and a password this should be set to 'auth'.
If all users are unauthenticated, the argument of this key *must* be 'public'.
If there are both kinds of users, this value has to be 'both'.

.br
.B 4. $config{'enable_domain_check'} -> [ 'true' | 'false' ]
.br
.P
By default, when authentication isn't used, postfilter checks whether the client's IP address has got
the right to post the current message. This is made checking how many attempts to post messages were
made by that IP address in the past and comparing the result with those rules that are set by the hash
and $config{'server_type'} is set to 'public' or $config{'server_type'} is set to 'both' and the
current message is sent by an authenticated user, postfilter makes a second check considering the
client's domain instead of his IP address. A client's domain is the reverse DNS of his IP address
without the last subdomain (so ip-45-32-135-21.someisp.tld becomes someisp.tld) and it's checked
against %public_rights_domain that is inside access.conf. This value should be always set to 'false' 
unless the server is attacked by many different IP addresses that are owned by a single ISP. This
statement has no effect for authenticated users; a value of 'false' makes useless %public_rights_domain

.br
.B 5. $config{'default_action_on_accept'} -> [ 'accept' | 'discard' | 'save' | 'reject' ]
.br
.P
This flag controls what postfilter has to do when a message passes all test. The most obvious setting is
changed by common users. If this is set to 'discard' Postfilter silently drops the post. The server gives 
back to the client the NNRP success code (240) but the article is discarded. If this setting is set to
pathspool/postfilter/saved/ without being accepted. With a value of 'reject' Potfilter rejects the article.
Every value except 'accept' will be probably interpreted by the users as a server defect because the
success code is returned to the client but the corresponding message doesn't become visibile in the 
destination groups.       
Note. Casual users don't need a value different than 'accept'. This feature is mostly useful for honeypot
servers and in order to study the spam aggressions.

.br
.B 6. $config{'default_action_on_reject'} -> [ 'accept' | 'discard' | 'save' | 'reject' ]
.br
.P
The postfilter behavior when an article fails the checks is set by this line. Possible values are the
same of the previous flag. Beware: if this setting is set to everything except 'reject', the client
will receive an NNRP success code instead an error code and this could encourage spammers to repeat
the abuses. 

.br
.B 7. $config{'public_user_id'} -> 'RegEX'
.br
.P
If $config{'server_type'} is set to 'both', postfilter has to determine which users are authenticated.
This key shows which INND userids describe *public* users. Those names are indicated inside readers.conf
in the 'default:' key. See readers.conf(5) for details.
Note: this is a regular expression (it's useful for large sites). This key isn't required if

.br
.B 8. $config{'reject_on_badwords_error'} -> [ 'true' | 'false' ]
.br
.P
If $config{'reject_on_badwords_error'} is set to 'true', if there's a syntax error inside badwords.conf
all messages are rejected with the error number 27 (Syntax error in badwords) or 38 (badwords.conf not
found). If $config{'reject_on_badwords_error'} is set to 'false' and there's an error, postfilter simply
skips the badwords check logging the error. The default value is set to 'true' due security reason.
On a side, a value of 'false' avoids that a syntax error inside badwords.conf or banlist.conf closes the
server since it isn't corrected. On the other side, if it's set to 'false', badwords and banlist checks
are disabled if a syntax error is found in the corresponding configuration files and this could open some
security hole if a server massively uses banlists.

.br
.B 9. $config{'reject_on_banlist_error'} -> [ 'true' | 'false' ]
.br
.P
This statement controls what postfilter has to do if $config{'use_banlist'} is set to 'true' and banlist.conf
includes some syntax error. If $config{'reject_on_banlist_error'} is set to 'true' and the banlist includes a
syntax error postfilter rejects ALL incoming articles with the error code 35 ('Syntax error in banlist file')
else postfilter skips the banlist. BEWARE: if this variable is set to 'true' and there's some syntax error in
the banlist file, *ALL* articles will be rejected; if it's set to 'false', syntax errors in banlist.conf
DISABLES the banlist check.

.br
.B 10. $config{'show_error_code'} -> [ 'true' | 'false' ]
.br
.P
This flag sets whether postfilter has to show in the NNRP response the internal error code. If this is set to
built using the @quickref array stored inside rules.conf. A value of 'false' makes hard for the sender to understand
what is wrong in his message. This setting doesn't affect system errors that are always reported in an explicit
manner in order to help the installation process.

.br
.B 11. $config{'period'} -> Positive, integer, number of seconds
.br
.P
Postfilter includes several time based barriers against users. This key shows how many seconds
postfilter has to go back when it verifies the past activities of each user.
A good value is 86400 (it means: 1 day).

.br
.B 12. $config{'short_period'} -> Positive, integer, number of seconds
.br
.P
Amount of time used by postfilter to check whether a flood is in progress. A good value for this is

.br
.B 13. $config{'trash_period'} -> Positive, integer, number of seconds
.br
.P
Postfilter records all attempts to send messages. This key sets how many seconds postfilter has
to keep these data inside its spool. This value must be greater or equal than $config{'period'}
A good value is 86400 (1 day), a long amount of time could make postfilter really slow.

.br
.B 14. $config{'enable_mysql'} -> [ 'true' | 'false' ]
.br
.P
If this is set to 'true', instead of reading its spool from a file, postfilter uses a mysql
database. If this feature is enabled, %mysql array (see below) must be properly filled with
the parameters for the mysql connection and the database must be created by hand (see doc/)

.br
.B 15. $config{'check_groups_existence'} -> [ 'true' | 'false' ]
.br
.P
If an user tries to send a message to a group that is not carried by the server, nnrpd rejects
it. If an user tries to send a message to an unexistent group and at the same time to a 
newsgroup that locally exists, it's needed to decide what postfilter has to do. If a value of
group that isn't locally available. This probably breaks RFC1036 but it's a considerable safer
behaviour since it prevents the users from posting into unknown and probably unwanted (notably
pedo) groups. A value of 'false' disables this check. This should be always set to 'true' if
a strict RFC1036 compliance isn't needed. 

.br
.B 16. $config{'check_users'} -> [ 'true' | 'false' ]
.br
.P
Postfilter is able to check whether an user has the right to post an article in a time. For
each IP or userid or domain, postfilter allows to set a maximum number of messages that can
be posted in two different amounts of time, an amount of identical copies of each article
that can be posted (multipost) and several other minor configuration settings. All this
parameters are inside rules.conf that must be configured before starting to seriously use
postfilter. This statement allows to set whether postfilter has to make this kind of checks:
a value of 'true' - which is the default and the reccomended choice - enables them, 'false'
disables any user check. 

.br
.B 17. $config{'check_tor'}    -> [ 'true' | 'false' ]
.br
.P
Tor is a network of anonymous proxies widely used by spammer and net-abusers. Postfilter is able
to check whether an article comes from a TOR proxy but this operation requires time since it's needed a
DNS query for each local IP address. A value of 'true' enables this check, 'false' disables it. 
It's higly recommended to keep this check active because TOR could become dangerous if TOR clients are
not controlled in some way. Those who make use of this feature must fill @localip (inside rules.conf) 
with *all* ip addresses used by he server.

.br
.B 18. $config{'tor_network'} -> [ 'allow' | 'reject' | 'mark' ]
.br
.P
If $config{'check_tor'} is set to 'true', postfilter checks whether each locally posted article comes
from a TOR proxy. This key sets what postfilter has to do when it detects an attempt to post an article
through a TOR proxy. If this is set to 'reject', all articles that come from that network are rejected. 
This is the safest value because TOR proxies are dangerous unless they're controlled. with 'allow', 
all messages that come from TOR will be accepted but a notice will be printed in the logs; 
if it is set to 'mark', articles are accepted but Path: is modified adding 'tor-network' and an extra 
header 'X-TOR-Router: $node_ip' is added.  

.br
.B 19. $config{'check_white-list'} -> [ 'true' | 'false' ]
.br
.P
If this variable is set to 'true', the white list is active. The hash %whitelist (see below) determines
which messages are automatically accepted with no checks. This structure has two elements: a key and a
regular expression that acts as argument. Every key is an header; if the regular expression matches the
contents of that header shown by the key, the message is accepted. Note that these articles are *not*
considered in the count of the accepted messages.

.br
.B 20. $config{'check_custom'} -> [ 'true' | 'false' ]
.br
.P
Even if postfilter is a general purpose filter, sometimes users could need to add their own code to the 
nnrpd filter. In order to allow this is a clear and safe manner, postfilter includes a file that is 
designed like a container for users' custom code (modules/custom.pm). This statement allows to choose
whether postfilter has to execute that file which includes (or could include) user's custom code. A
value of 'true' enables this, 'false' disables this kind of checks. Those who don't need to add their
code to postfilter can safety choose 'false' that makes postfilter a bit faster. Those who need to
customize postfilter must set this statement to 'true'.

.br
.B 21. $config{'check_banlist'} -> [ 'true' | 'false' ]
.br
.P
Whether to use the banlist is set by this key. More information about the postfilter banlist are stored
in the banlist.conf file.

.br
.B 22. $config{'score_banlist'} -> [ integer number ]
.br
.P
The maximum allowed score in the banlist (this is the default value, see the documentation of
banlist.conf)

.br
.B 23. $config{'check_rbl'}    -> [ 'true' | 'false' ]
.br
.P
Postfilter supports DNSBLs. If this parameter is set to 'true', the DNSBL listed in @dnsbl are checked
against the sender's ip address of each message. Please note that DNSBL check is slow because postfilter
has to check a single DNSBL per time.

.br
.B 24. $config{'check_uribl'}    -> [ 'true' | 'false' ]
.br
.P
Whether to check if an URL included in the body of each article is listed by SURBL or URIBL.com (that
are URIBLs, list that check whether 'urls' included in the body are know to be spam or scam

.br
.B 25. $config{'check_surbl'}    -> [ 'true' | 'false' ]
.br
.P
Whether to check if an URL included in the body of each article is listed by SURBL URIBL. This requires that

.br
.B 26. $config{'check_uriblcom'}    -> [ 'true' | 'false' ]
.br
.P
Whether to check if an URL included in the body of each article is listed by URIBL.com URIBL. This requires that

.br
.B 27. $config{'check_badwords'} -> [ 'true' | 'false' ]
.br
.P
Postfilter includes a (simple) spam dictionary which scans the body and the subject searching for defined
regular expression (see badwords.conf for details). 'true' activates it, 'false' keeps it inactive.

.br
.B 28. $config{'scan_body'} -> [ 'true' | 'false' ]
.br
.P
if $config{'use_badwords_scanner'} is set to 'true', this key determines if postfilter has to scan the
the body searching for spam words: 'true' enables this check, 'false' disables it.

.br
.B 29. $config{'max_score_on_body'} -> Integer, positive, number
.br
.P
For each spam word, a score is assigned. If the body matches a word, postfilter
sums this value to the global article score. If the body of an article exceeds the value of this key
as global score, the script rejects the message (see badwords.conf for details). This flag has effect
only if $config{'use_badwords_scanner'} and $config{'scan_body'} are set to 'true'.

.br
.B 30. $config{'scan_subject'} -> [ 'true' | 'false' ]
.br
.P
if $config{'use_badwords_scanner'} is set to 'true', this key determines if postfilter has to scan the
the subject searching for spam words: 'true' enables this check, 'false' disables it.

.br
.B 31. $config{'max_score_on_subject'} -> Integer, positive, number
.br
.P
if $config{'use_badwords_scanner'} and $config{'scan_subject'} are set to 'true', this flag determines
the maximum allowed score on the subject (see badwords.conf for details).

.br
.B 32. $config{'allow_control_cancel'} -> [ 'true' | 'false' ]
.br
.P
This key sets wheter the control cancel messages are allowed. The value of 'true' authorizes them,

.br
.B 33. $config{'allow_supersedes'} -> [ 'true' | 'false' ]
.br
.P
Whether an article can include Supersedes, Replaces, Cancels headers that replace an article with another

.br
.B 34. $config{'allow_uuencode'} -> [ 'true' | 'false' ]
.br
.P
If this is set to 'false' every article that seems to include some UUencoded text in the body is rejected.
This should be always set to 'false' if it isn't a binary news server.

.br

CHANGELOG
Postfilter 0.9.1

Enhancements:
	1. Postfilter requires also Encode::MIME::Header (postfilter.pl and postfilter-installer were updated)

Bugfixes:

	1. If the argument of From or Subject header includes a character that is forbidden, argument is converted to MIME format
	2. In order to check if a user has the right to post the message that is currently checked, postfilter now sums the values 
	   ​estracted from the current message to those related to messages already sent in the past. 
	3. Removed useless subroutines inside postfilter.pl
	4. If postfilter is configured in order to save accepted and/or refused local messages, each post is saved in a file inside
           [pathspool]/postfilter/saved/ using a name with format [3 digit error code].[UNIX timestamp] if it was rejected or 
	   00.[UNIX Timestamp] if it was accepted.

Postfilter 0.8.3

Enhancements:

	1. full support to INN 2.6
	2. A syntax checker for configuration files
	3. Installer allows to set innconfval path by hands
	4. Added $config{"allow_html"} that sets whether a message with HTML tags is rejected
	5. Distribution header is supported (Distribution: local|usenet)

Bugfixes:

	1. URIBLs work as expected
	2. Cache bugfix
	3. Installer works as expected
	4. Content-Type check now works correctly
	5. Invalid regexps in banlist.conf and badwords.conf now don't crash the filter
	6. Messages with non standard chars in the argument of an header are rejected
	


Postfilter 0.8.1

Enhancements:
	
	1. A better installer
	2. doc/ is now more complete
	3. postlegal is able to generate statistical data


Bugfixes:
 
       1. Postfilter works with INN 2.5	and 2.6








src/postfilter.pl
# Copyright (c) 2005-2019, Paolo Amoroso (Aioe) <freedom@aioe.org>
# All rights reserved.
# Postfilter version 0.9.1

#!/usr/bin/perl

use strict;

use Net::DNS;
use Date::Parse;
use Digest;
use Digest::MD5;
use Digest::SHA1;
use DBI;
use Encode::MIME::Header;

our (%hdr, $dbh, @access, $modify_headers, $body, $user, %config, %public_rights_ip, %public_rights_domain, %auth_rights, %ban_limits);
our (%headlist, @quickref, @saved_headers, %whitelist, @distributions, %mysql, %forbidden_crosspost, %scoreset, %maxscore);
our (@nomoderation, @htmlallowed, @htmltags, %extracontent, %forbidden_headers, @forbidden_groups, %dnsbl, @localip );

############################################################
#                                                          #
#                    Basic configuration                   #
#                                                          #
############################################################

my $use_innconfval	= "true";				# whether to use innconfval in order to determine the right path of
								# each file needed by postfilter. If this is set to "false", paths
								# are read from postingaccess.conf (that in this case is needed to
								# be properly configured).

my $innconfval 		= "/usr/lib/news/bin//innconfval";				# not needed if $use_innshellvar eq "false"
my $config_dir 		= "/etc/news/postfilter/"; 			# used only if $use_innshellvar eq "false"

my @files	= (						# configuration files that need to be loaded before analyzing each post
			"postfilter.conf",
			"rules.conf",
			"access.conf"
		  );
			
my @modules = (							# modules that *always* need to be loaded (they're in pathfilter/)
			"modules/access.pm",
			"modules/banchecks.pm",
			"modules/custom.pm",			# Custom rules must go here
			"modules/exit.pm",
			"modules/legal.pm",
			"modules/mysql.pm",
			"modules/other.pm",
			"modules/rbl.pm",
			"modules/style.pm",
			"modules/checkconfig.pm",
	      );


############################################################
#                                                          #
#                Do not edit before this line              #
#                                                          #
############################################################

########
#
# filter_post(): main subroutine
#
#######


sub filter_post()
{
        $modify_headers = 1;
        my $error_string;

#######################
# Server Status
#######################

        if ( $config{'server_status'} eq "closed" )
        {
		&log( "notice", "Server is closed due server_status is set to \"closed\""); 
                $error_string = &error( 48 );
                return $error_string;
        }
        elsif ( $config{'server_status'} eq "disabled" )
        {
                &log( "notice", "Filter is disabled, everything is accepted", "Message $hdr{'Message-ID'} accepted");
                return "";
        }



#######################
# Load Configurations 
#######################

        my $error_code = &load_config();
        if ( $error_code != 0 )
        {
		if ( $error_code == 42 )
		{
			return "Unable to run innconfval";
		}
		elsif ( $error_code == 43 )
		{
			return "Unable to load some configuration file";
		}
		elsif ( $error_code == 55 )
		{
			return "Unable to load some internal module";
		}
		else
		{
			&log( "err", "Strange load_config return value: $error_code" );
			return "Generic filter failure";
		}
        }


########################
# Syntax check
########################

	$error_code = &syntax_check();
        if ( $error_code != 0 )
        {
                $error_string = &error( $error_code );
                return $error_string;
        }

########################
# Filter Initialization
########################

        $error_code = &init_filter();
        if ( $error_code != 0 )
        {
                $error_string = &error( $error_code );
                return $error_string;
        }

#######################
# WhiteList Check
#######################

        if ( $config{'check_white-list'} eq "true" )
        {
               my $return_value = &check_whitelist();
               if ( $return_value != 0 )
               {
			$error_code = &whitelistchecks();	# return a numerical code

			if ( $error_code == 0 ) 		# Message accepted
			{
                        	$error_code = &success();
                        	if ( $error_code =~ /[0-9]+/ )                # errors
                        	{
                                	$error_string = &error( $error_code );
                                	return $error_string;
                        	}
				return $error_string;
			}
			else					# Message Rejected
			{
				my $error_string = &error( $error_code );
				return $error_string;
			}
               }
        }

#######################
# Forbidden Headers 
#######################

        $error_code = &style_filter();
        if ( $error_code != 0 )
        {
                $error_string = &error( $error_code );
                return $error_string;
        }



#######################
# RBLCheck  
#######################

	if ( $config{'check_rbl'} eq "true" )
	{
		$error_code = &rblcheck();
                if ( $error_code != 0 )
                {
                        $error_string = &error( $error_code );
                        return $error_string;
                }
	}


#######################
# URIBLCheck
#######################

        if ( $config{'check_uribl'} eq "true" )
        {
                $error_code = &check_uribl();
                if ( $error_code != 0 )
                {
                        $error_string = &error( $error_code );
                        return $error_string;
                }
        }



#######################
# TOR Check  
#######################

	if ( $config{'check_tor'} eq "true" )
	{
		$error_code = &check_tor();
		if ( $error_code != 0 )
        	{
                	$error_string = &error( $error_code );
                	return $error_string;
        	}
	}

#######################
# BanList  
#######################

        if ( $config{'check_banlist'} eq "true" )
        {
                $error_code = &check_banlist();
                if ( $error_code != 0 )
                {
                        $error_string = &error( $error_code );
                        return $error_string;
                }
        }

#######################
#  BadWords scanner
#######################

        if (  $config{'check_badwords'} eq "true" )
        {
                $error_code = &badwords();
                if ( $error_code != 0 )
                {
			if ( 
				( $error_code == 38 ) or 	# badwords.conf not found
				( $error_code == 27 )     	# syntax error inside badwords.conf
			   )
			{
				if ( $config{'reject_on_badwords_error'} ne "true" )
				{
					$error_string = &error( $error_code );
					&log( 'crit', $error_string );
				}	
				else
				{
					$error_string = &error( $error_code );
                        	        return $error_string;
				}
			}
			else
			{
	                	$error_string = &error( $error_code );
                        	return $error_string;
			}
                }
        }

#######################
# Check User Rights  
#######################

	if ( $config{'check_users'} eq "true" )
	{
        	$error_code = &check_user();
        	if ( $error_code != 0 )
        	{
                	$error_string = &error( $error_code );
                	return $error_string;
        	}
	}

#######################
# Custom rules
#######################

	if ( $config{'check_custom'} eq "true" )
	{
		$error_code = &custom_filter();
        	if ( $error_code != 0 )
        	{
                	$error_string = &error( $error_code );
                	return $error_string;
        	}
	}

#######################
# Modify Headers  
#######################

        $error_code = &mod_headers();
        if ( $error_code != 0 )
        {
                $error_string = &error( $error_code );
                return $error_string;
        }


#######################
# Log Data before accepting  
#######################

        $error_code = &success();
        if ( $error_code == 44 )		# unable to open legal.log
        {
                $error_string = &error( $error_code );
                return $error_string;
        }

	return $error_code;
	
}

########
#
# whitelistchecks(): Perform reduced checks for whitelisted users
#
#######


sub whitelistchecks()
{
	&log( "notice", "Message $hdr{'Message-ID'} matches a white list entry" );
 
	my $error_code;
       
	if ( $config{'use_banlist'} eq "true" )
        {
        	$error_code = &check_banlist();
                return $error_code if ( $error_code != 0 );
       	}
       	
	$error_code = &style_filter();
	return $error_code if ( $error_code != 0 );

	$error_code = &custom_filter(@access);
	return $error_code if ( $error_code != 0 );

        $error_code = &mod_headers();
	return $error_code if ( $error_code != 0 );
		
	return 0;

}

########
#
# load_config(): build paths and load configuration files and internal modules
#
#######


sub load_config
{	
	my %innshellvar;

#######################
# Run innconfval if needed
#######################

	if ( $use_innconfval eq "true" )
     	{
		my $syntax = $innconfval . " pathdb pathetc pathspool organization mailcmd pathfilter version";

		&log( "debug", "Executing $syntax" );
     		
		open my $SYNTAX, "$syntax|";
		if (!$SYNTAX)
		{
			&log( "err", "Unable to execute $syntax, aborting" );
			return 42;
		}

		my @elem = <$SYNTAX>;

		foreach (@elem)
		{
			chop $_;
			&log( "debug", "Server replies: $_" );
		}

		close $SYNTAX;

		$innshellvar{'file_active'}   	= $elem[0] . "/active";

		$config{'dir_config'} 		= $elem[1] . "/postfilter/";
                $innshellvar{'dir_config'} 	= $config{'dir_config'};
		$innshellvar{'file_badwords'} 	= $config{'dir_config'} . "/badwords.conf";
		$innshellvar{'file_banlist'} 	= $config{'dir_config'} . "/banlist.conf";

                $innshellvar{'file_access'}   	= $elem[2] . "/postfilter/access.log";
                $innshellvar{'file_legal'}    	= $elem[2] . "/postfilter/legal.log";
                $innshellvar{'dir_spool'}     	= $elem[2] . "/postfilter/";
		$innshellvar{'organization'}  	= $elem[3];
		$innshellvar{'sendmail'}      	= $elem[4];
		$innshellvar{'dir_filter'}      = $elem[5];
		$innshellvar{'version'}         = "2.4" if ( $elem[6] =~ /2\.4/ );
		$innshellvar{'version'}         = "2.5" if ( $elem[6] =~ /2\.5/ );
	}
	else
	{
		$config{'dir_config'} = $config_dir; 
	}

	my $tempdir = $config{'dir_config'};

#######################
# Load configuation files
#######################

	&log( "debug", "Loading configuration files from $tempdir" );
	foreach ( @files )
	{
                my $file = "$tempdir/$_";
                my $return;

                &log("debug", "Loading configuration file: $file");
                unless ($return = do $file)
                {
                        &log( "err", "Fatal Error: couldn't parse $file: $@" ) if $@;
                        &log( "err", "Fatal Error: couldn't do $file: $!" )    unless defined $return;
                        &log( "err", "Fatal Error: couldn't run $file" )       unless $return;
                        return 43;
                }
	}
	
	if ( $use_innconfval eq "true" )
	{
		foreach ( keys %innshellvar )		# this is made here in order to avoid that the values in postingaccess.conf
		{					# overwrite the values generated by innshellvar if $use_innconfval is "true"
			$config{$_} = $innshellvar{$_};
		}
	}

#######################
# Load internal modules
#######################

        &log( "debug", "Loading required internal perl modules" );
        foreach ( @modules )
        {
                my $file = "$config{'dir_filter'}/postfilter/$_";
                my $return;

                &log("debug", "Loading module: $file");
                unless ($return = do $file)
                {
                        &log( "err", "Fatal Error: couldn't parse $file: $@" ) if $@;
                        &log( "err", "Fatal Error: couldn't do $file: $!" )    unless defined $return;
                        &log( "err", "Fatal Error: couldn't run $file" )       unless $return;
                        return 55;
                }
        }

     	return 0;
}

########
#
# log(): Log data through nnrpd
#
#######

sub log()    # log( severity, string1, string2.... );
{
        my @strings = @_;

        my $severity = shift(@strings);

        foreach ( @strings )
        {
                INN::syslog( $severity, $_ );
        }
}





